### ISSUE TITLE
Rule proposal: Enforce hooks to be at the top of the test file

### ISSUE THREAD

**Issue by sindresorhus** [avajs/eslint-plugin-ava#196]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/196/badge.svg" />](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/196)

It's useful to be able to see the hooks first for readability.

Proposed order: `before`, `after`, `beforeEach`, `afterEach`, `test`

Thoughts?

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/196).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by novemberborn** [comment]
Is there an argument for `before`, `beforeEach`, `afterEach` and `after`?

What about the `.always` modifier?

**Comment by sindresorhus** [comment]
> Is there an argument for before, beforeEach, afterEach and after?

Yes, but I would prefer to sort them by semantic value, not when they are run. `before`/`after` are test-wide hooks so they are defined first, then test specific hooks, and then the actual tests. (This is also how we have it sorted in examples in the AVA readme).

> What about the .always modifier?

It should be after the hook with the same name without `.always`. So if you have `after.always` and `after`, `after` should come before `after.always`.

**Comment by novemberborn** [comment]
👍 let's do this.

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $60.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/49722492/issues/196) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/49722492/issues/196).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by GMartigny** [comment]
What would be the name of such a rule ? `hooks-order` ?

Hooks can have `.serial`, `.always`, `.cb` and `.skip` as modifiers, do we care for al ? what should be their effects ? their combined effects ?

Do you agree with this:

```js
// `serial` have priority 1 but before
test.serial.after()
test.serial.after.cb()
test.serial.after.always()
test.serial.after.always.cb()
test.serial.after.skip()
test.serial.after.cb.skip()
test.serial.after.always.skip()
test.serial.after.always.cb.skip()

test.after()

test.after.cb()
test.after.always()
test.after.always.cb()
// `skip` have priority 2
test.after.skip()
test.after.cb.skip()
// `always` have priority 3
test.after.always.skip()
test.after.always.cb.skip() // cb is the least impactful modifier

test()
```

Other possibility is to prioritize by position on the call `test.serial.after.always.cb.skip()` (so 1st `serial`, `always`, `cb` then `skip`)

**Comment by novemberborn** [comment]
I don't think the concurrency (`.serial`) or execution method (`.cb`) are relevant here. AVA already orchestrates those and it's very test suite specific.
